runtime/ffi/ffi.cpp

```diff
@@ -0,0 +1,200 @@
+  TAG_TYPE(sint32)
+  TAG_TYPE(uint64)
+  TAG_TYPE(sint64)
+  TAG_TYPE(float)
+  TAG_TYPE(double)
+  TAG_TYPE(uchar)
+  TAG_TYPE(schar)
+  TAG_TYPE(ushort)
+  TAG_TYPE(sshort)
+  TAG_TYPE(uint)
+  TAG_TYPE(sint)
+  TAG_TYPE(ulong)
+  TAG_TYPE(slong)
+  TAG_TYPE(longdouble)
+
+  mpz_ptr move_int(mpz_t);
+  char * getTerminatedString(string * str);
+
+  mpz_ptr hook_LIST_size(struct list * l);
+  block * hook_LIST_get(struct list * l, int idx);
+
+  static void * so_lib_handle() {
+    static void * handle = NULL;
+
+    if (handle == NULL) {
+      handle = dlopen(NULL, RTLD_LAZY);
+    }
+
+    return handle;
+  }
+
+  static ffi_type getTypeFromSymbol(uint64_t symbol) {
+    if (symbol == tag_type_void()) {
+      return ffi_type_void;
+    } else if (symbol == tag_type_uint8()) {
+      return ffi_type_uint8;
+    } else if (symbol == tag_type_sint8()) {
+      return ffi_type_sint8;
+    } else if (symbol == tag_type_uint16()) {
+      return ffi_type_uint16;
+    } else if (symbol == tag_type_sint16()) {
+      return ffi_type_sint16;
+    } else if (symbol == tag_type_uint32()) {
+      return ffi_type_uint32;
+    } else if (symbol == tag_type_sint32()) {
+      return ffi_type_sint32;
+    } else if (symbol == tag_type_uint64()) {
+      return ffi_type_uint64;
+    } else if (symbol == tag_type_sint64()) {
+      return ffi_type_sint64;
+    } else if (symbol == tag_type_float()) {
+      return ffi_type_float;
+    } else if (symbol == tag_type_double()) {
+      return ffi_type_double;
+    } else if (symbol == tag_type_uchar()) {
+      return ffi_type_uchar;
+    } else if (symbol == tag_type_schar()) {
+      return ffi_type_schar;
+    } else if (symbol == tag_type_ushort()) {
+      return ffi_type_ushort;
+    } else if (symbol == tag_type_sshort()) {
+      return ffi_type_sshort;
+    } else if (symbol == tag_type_uint()) {
+      return ffi_type_uint;
+    } else if (symbol == tag_type_sint()) {
+      return ffi_type_sint;
+    } else if (symbol == tag_type_ulong()) {
+      return ffi_type_ulong;
+    } else if (symbol == tag_type_slong()) {
+      return ffi_type_slong;
+    } else if (symbol == tag_type_longdouble()) {
+      return ffi_type_longdouble;
+    } else {
+      throw std::invalid_argument("Arg is not a supported type");
+    }
+  }
+
+  block * hook_FFI_call(mpz_t addr, struct list * args, struct list * types, block * ret) {
+    ffi_cif cif;
+    ffi_type *argtypes, rtype;
+    void (* address)();
+
+    if (!mpz_fits_ulong_p(addr)) {
+      throw std::invalid_argument("Addr is too large");
+    }
+    *(void **)(&address) = (void *) mpz_get_ui(addr);
+
+    mpz_ptr nargs_ptr = hook_LIST_size(args);
+    mpz_ptr ntypes_ptr = hook_LIST_size(types);
+    if (!mpz_fits_ulong_p(nargs_ptr)) {
+      throw std::invalid_argument("Args list is too large");
+    }
+    if (!mpz_fits_ulong_p(ntypes_ptr)) {
+      throw std::invalid_argument("Types list is too large");
+    }
+
+    size_t nargs = mpz_get_si(nargs_ptr);
```

Comment:
this will be removed by the above comment, but note that size_t is unsigned so it really should have been mpz_get_ui.